yarn-project/noir-contracts/src/contracts/token_bridge_contract/src/main.nr

```diff
@@ -0,0 +1,172 @@
+
+        1
+    }
+
+    // Burns the appropriate amount of tokens and creates a L2 to L1 withdraw message publicly
+    // Requires `msg.sender` to give approval to the bridge to burn tokens on their behalf using witness signatures 
+    #[aztec(public)]
+    fn exit_to_l1_public(
+        recipient: EthereumAddress, // ethereum address to withdraw to
+        amount: Field,
+        callerOnL1: EthereumAddress, // ethereum address that can call this function on the L1 portal (0x0 if anyone can call)
+        nonce: Field, // nonce used in the approval message by `msg.sender` to let bridge burn their tokens on L2
+    ) -> Field {
+        let storage = Storage::init(Context::public(&mut context));
+
+        // Send an L2 to L1 message
+        let content = get_withdraw_content_hash(recipient.address, amount, callerOnL1.address);
+        context.message_portal(content);     
+
+        // Burn tokens 
+        Token::at(storage.token.read()).burn_public(context, context.msg_sender(), amount, nonce);
+   
+        1
+    }
+
+    // Consumes a L1->L2 message and calls the token contract to mint the appropriate amount in private assets
+    // User needs to call token.redeem_shield() to get the private assets
+    #[aztec(private)]
+    fn claim_private(
+        amount: Field,
+        secret_hash_for_redeeming_minted_notes: Field, // secret hash used to redeem minted notes at a later time. This enables anyone to call this function and mint tokens to a user on their behalf
+        canceller: EthereumAddress,
+        msg_key: Field, // L1 to L2 message key as derived from the inbox contract
+        secret_for_L1_to_L2_message_consumption: Field, // secret used to consume the L1 to L2 message
+    ) -> Field {
+        // Consume L1 to L2 message and emit nullifier
+        let content_hash = get_mint_private_content_hash(amount, secret_hash_for_redeeming_minted_notes, canceller.address);
+        context.consume_l1_to_l2_message(msg_key, content_hash, secret_for_L1_to_L2_message_consumption);
+
+        // Mint tokens on L2 
+        // `mint_private` on token is public. So we call an internal public function 
+        // which then calls the public method on the token contract.
+        // Since the secret_hash is passed, no secret is leaked. 
+        context.call_public_function(
+            context.this_address(),
+            compute_selector("_call_mint_on_token(Field,Field)"),
+            [amount, secret_hash_for_redeeming_minted_notes],
+        );
+
+        1
+    }
+
+    // Burns the appropriate amount of tokens and creates a L2 to L1 withdraw message privately
+    // Requires `msg.sender` (caller of the method) to give approval to the bridge to burn tokens on their behalf using witness signatures 
+    #[aztec(private)]
+    fn exit_to_l1_private(
+        recipient: EthereumAddress, // ethereum address to withdraw to
+        token: AztecAddress,
```

Comment:
The token here is only really used for validation in here while the rest are passed into the hash. So I would prefer that we had that closer to the end, such that the tree values put into the hash are in a row, but practically don't matter. Should probably get a style-guide going for this kind of thing.